[GTK][WPE][Skia] Add canvas 2D operation recording for batched replay

nikolaszimmermann opened the pull request:
#### 05c4585aef2819a1881b7d910e70ae55492e0282
<pre>
[GTK][WPE][Skia] Add canvas 2D operation recording for batched replay
<a href="https://bugs.webkit.org/show_bug.cgi?id=303575">https://bugs.webkit.org/show_bug.cgi?id=303575</a>

Reviewed by Carlos Garcia Campos.

Introduce canvas 2D operation recording for GPU-accelerated canvas rendering.
Instead of executing drawing commands immediately, operations are recorded into
an SkPicture and replayed in batch when the canvas contents are actually needed.
This reduces GPU state change overhead for workloads with many small drawing
operations.

When a canvas ImageBuffer is first accessed, we create a recording context backed
by an SkPictureRecorder. All subsequent drawing operations are captured into this
recording rather than being executed on the GPU surface. This continues until a
&quot;materialization point&quot; is reached - a point where the actual pixel contents are
needed. The optimal case is prepareForDisplay(), where the recording is replayed
just before compositing. Other materialization points include createNativeImageReference()
for snapshots, getPixelBuffer()/putPixelBuffer() for CPU access, and flushContext()
for explicit synchronization.

At the materialization point, the recording is finalized and replayed to the GPU
surface. The drawing state (fill/stroke colors, composite modes, transforms, etc.)
is preserved across this transition. After flushing, the canvas switches to direct
rendering mode for all subsequent operations - we don&apos;t restart recording, as the
batching benefit has already been realized for the initial burst of drawing commands.

A SkiaSwitchableCanvas (derived from SkNWayCanvas) serves as the stable canvas
reference held by GraphicsContextSkia, allowing us to transparently redirect
drawing operations from the recording canvas to the surface canvas after flushing.

Cross-thread transfer (e.g., for OffscreenCanvas) is handled by flushing any
pending recording on the source thread before transfer, since Skia surfaces are
not thread-safe. The destination thread simply waits for the GPU fence rather
than recreating the surface. Additionally, when drawing GPU-backed images whose
textures were created in a different GL context, we rewrap them using
SkImages::BorrowTextureFrom() to make them valid for the current context.

Covered by existing tests.

* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.h:
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.cpp:
(WebCore::ImageBufferSkiaAcceleratedBackend::~ImageBufferSkiaAcceleratedBackend):
(WebCore::ImageBufferSkiaAcceleratedBackend::context):
(WebCore::ImageBufferSkiaAcceleratedBackend::ensureCanvasRecordingContext):
(WebCore::ImageBufferSkiaAcceleratedBackend::copyGraphicsState):
(WebCore::ImageBufferSkiaAcceleratedBackend::flushCanvasRecordingContextIfNeeded):
(WebCore::ImageBufferSkiaAcceleratedBackend::flushContext):
(WebCore::ImageBufferSkiaAcceleratedBackend::prepareForDisplay):
(WebCore::ImageBufferSkiaAcceleratedBackend::createNativeImageReference):
(WebCore::ImageBufferSkiaAcceleratedBackend::getPixelBuffer):
(WebCore::ImageBufferSkiaAcceleratedBackend::putPixelBuffer):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.h:
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaSurfaceBackend.h:
* Source/WebCore/platform/graphics/skia/SkiaRecordingResult.cpp:
* Source/WebCore/platform/graphics/skia/SkiaRecordingResult.h:
* Source/WebCore/platform/graphics/skia/SkiaReplayCanvas.cpp:
* Source/WebCore/platform/graphics/skia/SkiaReplayCanvas.h:

Canonical link: <a href="https://commits.webkit.org/305273@main">https://commits.webkit.org/305273@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/57c6db3fe881a13037e3c89bdc3f08c7f577ff1d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/137991 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/10355 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/49370 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/146059 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/90966 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/139864 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/11057 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/10502 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/105527 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/90966 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/140936 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/8249 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/123712 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/86378 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/7863 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/5617 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/6342 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/117259 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/41888 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/148770 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/10038 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/42447 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/113931 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/10055 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/8474 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/114262 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/29025 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/7802 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/119975 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/64762 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/10084 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/37957 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/9815 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/73652 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/10025 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/9876 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -296,21 +297,36 @@ void GraphicsContextSkia::drawNativeImage(NativeImage& nativeImage, const FloatR
         });
     } else
-        trackAcceleratedRenderingFenceIfNeeded(image);
+        trackAcceleratedRenderingFenceIfNeeded(rewrappedImage ? rewrappedImage : image);
```
maybe we need a comment or better names, but I find a bit confusing this function now. We have image, useImage, rasterImage and rewrappedImage and I don't understand what they are to be honest.

nikolaszimmermann replied:
In the case of a GPU-backed `image` coming from another thread, it gets "re-wrapped" in the new GL/GR context, so that it's accessible. If that is the case one is not allowed to sample from `image`, but has to use `rewrappedImage`.

`rasterImage` is a special case, where we have to do a GPU -> CPU conversion for drop shadows, due to a Skia limitation that we have to workaround (if we don't do that, the image is just empty).

`useImage` is really confusing.... will think about it.

nikolaszimmermann replied:
Ok, here's my proposal:

```
diff --git a/Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp b/Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp
index 57c94ada446d..89f4edf1b4db 100644
--- a/Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp
+++ b/Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp
@@ -295,9 +295,10 @@ void GraphicsContextSkia::drawNativeImage(NativeImage& nativeImage, const FloatR
     bool inExtraTransparencyLayer = false;
     auto clampingConstraint = options.strictImageClamping() == StrictImageClamping::Yes ? SkCanvas::kStrict_SrcRectConstraint : SkCanvas::kFast_SrcRectConstraint;
 
-    SkImage* useImage = image.get();
+    // 'imageInThisThread' is either the incoming 'image', or a wrapper around the 'image' accessible in the current thread.
+    // When you want to access the image, use 'imageInThisThread' if it's non-zero or 'image'.
+    sk_sp<SkImage> imageInThisThread;
 
-    sk_sp<SkImage> rewrappedImage;
     if (image->isTextureBacked()) {
         auto* grContext = PlatformDisplay::sharedDisplay().skiaGrContext();
 
@@ -305,30 +306,33 @@ void GraphicsContextSkia::drawNativeImage(NativeImage& nativeImage, const FloatR
         // If the image was created in a different thread/context, we need to rewrap it.
         if (grContext && !image->isValid(grContext->asRecorder())) {
             GrBackendTexture backendTexture;
-            if (SkImages::GetBackendTextureFromImage(image.get(), &backendTexture, false)) {
-                if ((rewrappedImage = SkImages::BorrowTextureFrom(grContext, backendTexture, kTopLeft_GrSurfaceOrigin, image->colorType(), image->alphaType(), image->refColorSpace())))
-                    useImage = rewrappedImage.get();
-            }
+            if (SkImages::GetBackendTextureFromImage(image.get(), &backendTexture, false))
+                imageInThisThread = SkImages::BorrowTextureFrom(grContext, backendTexture, kTopLeft_GrSurfaceOrigin, image->colorType(), image->alphaType(), image->refColorSpace());
         }
     }
 
-    sk_sp<SkImage> rasterImage;
+    // 'imageForDrawing' references the incoming image (or if it's not directly accessible in this thread the 'imageInThisThread' which is).
+    // However, if we have to make a raster copy (see the hasDropShadow() case below), then imageForDrawing will point to the raster copy instead.
+    // This is the image we have to pass on to m_canvas.drawImageRect(...) below.
+    SkImage* imageForDrawing = imageInThisThread ? imageInThisThread.get() : image.get();
+
+    sk_sp<SkImage> imageRasterCopy;
     if (hasDropShadow()) {
-        if (image->isTextureBacked()) {
+        if (imageInThisThread->isTextureBacked()) {
             if (renderingMode() == RenderingMode::Unaccelerated) {
                 // When drawing GPU-backed image on CPU-backed canvas with filter, we need to convert image to CPU-backed one.
-                rasterImage = useImage->makeRasterImage();
-                useImage = rasterImage.get();
+                imageRasterCopy = imageInThisThread ? imageInThisThread->makeRasterImage() : image->makeRasterImage();
+                imageForDrawing = imageRasterCopy.get();
             } else
-                trackAcceleratedRenderingFenceIfNeeded(rewrappedImage ? rewrappedImage : image);
+                trackAcceleratedRenderingFenceIfNeeded(imageInThisThread ? imageInThisThread : image);
         }
         inExtraTransparencyLayer = drawOutsetShadow(paint, [&](const SkPaint& paint) {
-            m_canvas.drawImageRect(useImage, normalizedSrcRect, normalizedDestRect, toSkSamplingOptions(m_state.imageInterpolationQuality()), &paint, clampingConstraint);
+            m_canvas.drawImageRect(imageForDrawing, normalizedSrcRect, normalizedDestRect, toSkSamplingOptions(m_state.imageInterpolationQuality()), &paint, clampingConstraint);
         });
     } else
-        trackAcceleratedRenderingFenceIfNeeded(rewrappedImage ? rewrappedImage : image);
+        trackAcceleratedRenderingFenceIfNeeded(imageInThisThread ? imageInThisThread : image);
 
-    m_canvas.drawImageRect(useImage, normalizedSrcRect, normalizedDestRect, toSkSamplingOptions(m_state.imageInterpolationQuality()), &paint, clampingConstraint);
+    m_canvas.drawImageRect(imageForDrawing, normalizedSrcRect, normalizedDestRect, toSkSamplingOptions(m_state.imageInterpolationQuality()), &paint, clampingConstraint);
     if (inExtraTransparencyLayer)
         endTransparencyLayer();
```

nikolaszimmermann replied:
Better comments, better names, hopefully. Let's see what you say :-)

carlosgcampos commented on Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.cpp:
```diff
@@ -123,20 +146,168 @@ ImageBufferSkiaAcceleratedBackend::ImageBufferSkiaAcceleratedBackend(const Param
+
+    // Only flush if there's actually an active recording to flush.
+    if (!m_canvasRecordingContext || !m_shouldUseCanvasRecording)
+        return;
```
This is already done by flushCanvasRecordingContextIfNeeded(), no?

nikolaszimmermann replied:
Good catch.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.cpp:
```diff
@@ -51,6 +55,25 @@ WTF_IGNORE_WARNINGS_IN_THIRD_PARTY_CODE_END
+class SkiaSwitchableCanvas final : public SkNWayCanvas {
+WTF_MAKE_TZONE_ALLOCATED(SkiaSwitchableCanvas);
+public:
+    SkiaSwitchableCanvas(int width, int height)
```
```suggestion
    explicit SkiaSwitchableCanvas(const IntSize&)
```

carlosgcampos replied:
This way you can just pass size() when creating this.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.cpp:
```diff
@@ -123,10 +146,131 @@ ImageBufferSkiaAcceleratedBackend::ImageBufferSkiaAcceleratedBackend(const Param
+    m_canvasRecordingContext->beginRecording();
+}
+
+void ImageBufferSkiaAcceleratedBackend::copyGraphicsState(GraphicsContextSkia& from, GraphicsContextSkia& to)
```
```suggestion
void ImageBufferSkiaAcceleratedBackend::copyGraphicsState(const GraphicsContextSkia& from, GraphicsContextSkia& to)
```

carlosgcampos commented on Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.cpp:
```diff
@@ -123,10 +146,131 @@ ImageBufferSkiaAcceleratedBackend::ImageBufferSkiaAcceleratedBackend(const Param
+
+void ImageBufferSkiaAcceleratedBackend::copyGraphicsState(GraphicsContextSkia& from, GraphicsContextSkia& to)
+{
+    auto& fromState = from.state();
```
```suggestion
    const auto& fromState = from.state();
```

carlosgcampos commented on Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.cpp:
```diff
@@ -123,10 +146,131 @@ ImageBufferSkiaAcceleratedBackend::ImageBufferSkiaAcceleratedBackend(const Param
+        if (auto fence = flushCanvasRecordingContextIfNeeded())
+            fence->serverWait();
+        return;
+    }
```
The recording context is only created when purpose is canvas, so I don't think we need to check it here again.

nikolaszimmermann committed:
[GTK][WPE][Skia] Add canvas 2D operation recording for batched replay
https://bugs.webkit.org/show_bug.cgi?id=303575

Reviewed by Carlos Garcia Campos.

Introduce canvas 2D operation recording for GPU-accelerated canvas rendering.
Instead of executing drawing commands immediately, operations are recorded into
an SkPicture and replayed in batch when the canvas contents are actually needed.
This reduces GPU state change overhead for workloads with many small drawing
operations.

When a canvas ImageBuffer is first accessed, we create a recording context backed
by an SkPictureRecorder. All subsequent drawing operations are captured into this
recording rather than being executed on the GPU surface. This continues until a
"materialization point" is reached - a point where the actual pixel contents are
needed. The optimal case is prepareForDisplay(), where the recording is replayed
just before compositing. Other materialization points include createNativeImageReference()
for snapshots, getPixelBuffer()/putPixelBuffer() for CPU access, and flushContext()
for explicit synchronization.

At the materialization point, the recording is finalized and replayed to the GPU
surface. The drawing state (fill/stroke colors, composite modes, transforms, etc.)
is preserved across this transition. After flushing, the canvas switches to direct
rendering mode for all subsequent operations - we don't restart recording, as the
batching benefit has already been realized for the initial burst of drawing commands.

A SkiaSwitchableCanvas (derived from SkNWayCanvas) serves as the stable canvas
reference held by GraphicsContextSkia, allowing us to transparently redirect
drawing operations from the recording canvas to the surface canvas after flushing.

Cross-thread transfer (e.g., for OffscreenCanvas) is handled by flushing any
pending recording on the source thread before transfer, since Skia surfaces are
not thread-safe. The destination thread simply waits for the GPU fence rather
than recreating the surface. Additionally, when drawing GPU-backed images whose
textures were created in a different GL context, we rewrap them using
SkImages::BorrowTextureFrom() to make them valid for the current context.

Covered by existing tests.

* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.h:
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.cpp:
(WebCore::ImageBufferSkiaAcceleratedBackend::~ImageBufferSkiaAcceleratedBackend):
(WebCore::ImageBufferSkiaAcceleratedBackend::context):
(WebCore::ImageBufferSkiaAcceleratedBackend::ensureCanvasRecordingContext):
(WebCore::ImageBufferSkiaAcceleratedBackend::copyGraphicsState):
(WebCore::ImageBufferSkiaAcceleratedBackend::flushCanvasRecordingContextIfNeeded):
(WebCore::ImageBufferSkiaAcceleratedBackend::flushContext):
(WebCore::ImageBufferSkiaAcceleratedBackend::prepareForDisplay):
(WebCore::ImageBufferSkiaAcceleratedBackend::createNativeImageReference):
(WebCore::ImageBufferSkiaAcceleratedBackend::getPixelBuffer):
(WebCore::ImageBufferSkiaAcceleratedBackend::putPixelBuffer):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.h:
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaSurfaceBackend.h:
* Source/WebCore/platform/graphics/skia/SkiaRecordingResult.cpp:
* Source/WebCore/platform/graphics/skia/SkiaRecordingResult.h:
* Source/WebCore/platform/graphics/skia/SkiaReplayCanvas.cpp:
* Source/WebCore/platform/graphics/skia/SkiaReplayCanvas.h:

Canonical link: https://commits.webkit.org/305273@main
